### PR TITLE
Implement Persistent Workers as an execution path

### PR DESCRIPTION
**Pull request by wiwa**
Followup to #1195 

Add a new execution pathway in `worker/Executor.java` to use persistent workers via `PersistentExecutor`, like `DockerExecutor`.

Mostly unchanged from the form we used to experiment back at Twitter, but now with tests.

Co-authored-by: Shane Delmore [shane@delmore.io](mailto:shane@delmore.io)

### CODE REVIEW

**Comment by werkt** [comment]
@wiwa any chance this will get updated? I'd love to see this in, and could pick up the mantle if you don't have any time for it.

**Comment by wiwa** [comment]
@werkt  Sorry, didn't see your comment! I'm mostly worried about test cases for this. I'll take a look this weekend again. Maybe I should just add a basic test case first and start from there.

**Comment by ShaneDelmore** [comment]
@werkt We added some tests today, will come back and continue adding more tests in the following weeks.

File: src/main/java/build/buildfarm/worker/persistent/ProtoCoordinator.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        .append(">:\ngetOutputPathsList:\n")
+        .append(context.outputPaths)
+        .append("getOutputFilesList:\n")
+        .append(context.outputFiles)
+        .append("getOutputDirectoriesList:\n")
+        .append(context.outputDirectories);
+
+    logger.severe(sb.toString());
+
+    e.printStackTrace();
```
Comment (werkt) [inline src/main/java/build/buildfarm/worker/persistent/ProtoCoordinator.java:193]: send this to the logger with `log` method

File: src/main/java/build/buildfarm/worker/persistent/WorkerInputs.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.worker.WorkerProtocol.Input;
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.logging.Logger;
+
+public class WorkerInputs {
+  private static final Logger logger = Logger.getLogger(WorkerInputs.class.getName());
```
Comment (werkt) [inline src/main/java/build/buildfarm/worker/persistent/WorkerInputs.java:13]: `@Log` this

File: src/main/java/build/buildfarm/worker/persistent/WorkerInputs.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            + "\nallInputs: "
+            + pathInputs.keySet()
+            + "\ntoolInputs: "
+            + toolInputs
+            + "\nabsToolInputs: "
+            + absToolInputs;
+
+    logger.fine(inputsDebugMsg);
+
+    System.out.println(inputsDebugMsg);
```
Comment (werkt) [inline src/main/java/build/buildfarm/worker/persistent/WorkerInputs.java:102]: debug print?

File: src/main/java/build/buildfarm/worker/util/BUILD [hunk 4]
```diff
@@ -0,0 +1,24 @@
+java_library(
+    name = "util",
+    srcs = glob(["*.java"]),
+    visibility = ["//visibility:public"],
+    deps = [
+        "//persistentworkers/src/main/protobuf:worker_protocol_java_proto",
```
Comment (werkt) [inline src/main/java/build/buildfarm/worker/util/BUILD:6]: alpha

File: src/test/java/build/buildfarm/instance/shard/JedisCasWorkerMapTest.java [hunk 5]
```diff
@@ -20,7 +20,6 @@
 
 @RunWith(JUnit4.class)
 public class JedisCasWorkerMapTest {
-
```
Comment (werkt) [inline src/test/java/build/buildfarm/instance/shard/JedisCasWorkerMapTest.java:23]: ws

File: src/test/java/build/buildfarm/worker/persistent/BUILD [hunk 6]
```diff
@@ -0,0 +1,36 @@
+java_test(
+    name = "tests",
+    size = "small",
+    srcs = glob(["*.java"]),
+    test_class = "build.buildfarm.AllTests",
+    deps = [
+        "//persistentworkers/src/main/java/persistent/bazel:bazel-persistent-workers",
+        "//persistentworkers/src/main/java/persistent/common:persistent-common",
+        "//persistentworkers/src/main/java/persistent/common/util",
+        "//persistentworkers/src/main/protobuf:worker_protocol_java_proto",
```
Comment (werkt) [inline src/test/java/build/buildfarm/worker/persistent/BUILD:10]: alpha

**Review by werkt — CHANGES_REQUESTED** [review]
New files need some copyrights, and a buildifier run should fix most of it.

File: src/main/java/build/buildfarm/worker/BUILD [hunk 7]
```diff
@@ -4,12 +4,16 @@ java_library(
     plugins = ["//src/main/java/build/buildfarm/common:lombok"],
     visibility = ["//visibility:public"],
     deps = [
+        "//persistentworkers/src/main/java/persistent/bazel:bazel-persistent-workers",
+        "//persistentworkers/src/main/java/persistent/common:persistent-common",
```
[the remark this answers is not in the document]
Comment (wiwa) [reply src/main/java/build/buildfarm/worker/BUILD:8]: These are in fact sorted already (`//persistentworkers/...` before `//src/...`).

File: src/main/java/build/buildfarm/common/config/DequeueMatchSettings.java [hunk 8]
```diff
@@ -9,7 +9,6 @@
 
 @Data
 public class DequeueMatchSettings {
-
```
[the remark this answers is not in the document]
Comment (wiwa) [reply src/main/java/build/buildfarm/common/config/DequeueMatchSettings.java:12]: Unfortunately, this ws change (and the other one) are both from running `./.bazelci/format.sh` (on my WSL box).

File: src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+      return Code.INVALID_ARGUMENT;
+    }
+
+    // TODO revisit why this was necessary in the first place
+    ImmutableMap<String, String> env;
+    if (executionName.equals(JAVAC_EXEC_NAME)) {
+      env = ImmutableMap.of();
+    } else {
+      env = envVars;
+    }
```
[the remark this answers is not in the document]
Comment (wiwa) [reply src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java:98]: I'll put a comment (and "TODO" lol) up top:
```
// I believe the reason has to do with JavaBuilder workers not relying on env vars,
// as compared to rules_scala, only reading info from the argslist of each command.
// That would mean the Java worker keys should be invariant to the env vars we see.
```

File: src/main/java/build/buildfarm/worker/persistent/BUILD [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    srcs = glob(["*.java"]),
+    visibility = ["//visibility:public"],
+    deps = [
+        "//persistentworkers/src/main/java/persistent/bazel:bazel-persistent-workers",
+        "//persistentworkers/src/main/java/persistent/common:persistent-common",
+        "//persistentworkers/src/main/java/persistent/common/util",
+        "//persistentworkers/src/main/protobuf:worker_protocol_java_proto",
+        "//src/main/java/build/buildfarm/common",
+        "//src/main/java/build/buildfarm/worker/resources",
+        "//src/main/java/build/buildfarm/worker/util",
```
[the remark this answers is not in the document]
Comment (wiwa) [reply src/main/java/build/buildfarm/worker/persistent/BUILD:13]: sorted

**Comment by wiwa** [comment]
@ShaneDelmore I was testing this out with the current `bazel` (`6.4.0`) which now supports the `--experimental_remote_mark_tool_inputs` flag. It turns out we'll need to add `persistentWorkerKey` to the worker config (`config/Worker.java`), specifically allowing `DequeueMatchSettings` to accept `persistentWorkerKey`, which used to work as just a platform property `persistentWorkerKey: *`. We also need to figure out how to pass the worker init command from `bazel` (i.e. `persistentWorkerCommand`) -- this should be a small modification to `bazel` itself. Instead, I tested with taking a command like `java <flags> JavaBuilder.jar <@argfiles>` and truncating the `@argsfiles` and before adding the `--persistent_worker` flag. However, this assumes that the first argument(s) of the command is an `@argsfile`, which I'm not sure is true in general, and is probably why we ended up with a `persistentWorkerCommand` requirement. Otherwise, it seems to work (i.e. `cmp -l <build-with-pw-repo>/bazel-bin/src/main/java/build/buildfarm/buildfarm-server_deploy.jar <build-with-local-repo>/bazel-bin/src/main/java/build/buildfarm/buildfarm-server_deploy.jar`).

File: src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (werkt) [reply src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java:114]: I'm going to hope that the lack of JavaBuilder support means that nobody is going to expect their execution to change with any env vars across instantiation. Probably a big hope. Also its not immediately clear that this `env` only has ties to the key creation (I figured it had to do with the execution overall until looking with your update where this applies). Maybe move it down under 'Make Key' and label it as `keyEnv`?

File: src/main/java/build/buildfarm/worker/BUILD (hunk 7 above)
[the remark this answers is not in the document]
Comment (werkt) [reply src/main/java/build/buildfarm/worker/BUILD:8]: I'm blind, thanks for the gut check!

File: src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (wiwa) [reply src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java:114]: It is in fact tied to the execution as well as key creation. I think the reality is that these lines of code should be updated in the future after further investigation and/or bug reports. However, this was what was tested back at Twitter and from what I could tell, it was required for correct functionality. This shouldn't be taken as "lack of JavaBuilder support", but rather, this *is* JavaBuilder support.

[the remark this answers is not in the document]
Comment (werkt) [reply src/main/java/build/buildfarm/worker/persistent/PersistentExecutor.java:114]: What I meant on 'lack of JavaBuilder support' there was its lack of interpretation of env vars to affect its runtime - sorry if that was confusing.

And I'm not seeing the use of the env anywhere except in the call to Keymaker.make, but if you're saying that it is used somehow through injection where it affects the execution by way of the key constituency, sounds good.

**Review by werkt — APPROVED** [review]

**Pull request merged**